adapters/repos/db/snapshot.go

```diff
@@ -0,0 +1,96 @@
+package db
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/semi-technologies/weaviate/entities/snapshots"
+)
+
+// CreateSnapshot creates a new active snapshot for all state in this index across
+// all its shards. It is safe to copy any file referenced in the snapshot, as the
+// active state in the snapshot guarantees that those files cannot be modified.
+//
+// There can only be one active snapshot at a time, and creating a snapshot will
+// fail on any Index that already has an active snapshot.
+//
+// Make sure to call ReleaseSnapshot for this snapshot's ID once you have finished
+// copying the files to make sure background and maintenance processes can resume.
+func (i *Index) CreateSnapshot(ctx context.Context, id string) (*snapshots.Snapshot, error) {
+	if err := i.initSnapshot(id); err != nil {
+		return nil, err
+	}
+
+	snap := snapshots.New(id, time.Now(), i.Config.RootPath)
+
+	for _, shard := range i.Shards {
+		if err := shard.createSnapshot(ctx, snap); err != nil {
+			return nil, err
+		}
+	}
+
+	shardingState, err := i.marshalShardingState()
+	if err != nil {
+		return nil, errors.Wrap(err, "create snapshot")
+	}
+
+	schema, err := i.marshalSchema()
+	if err != nil {
+		return nil, errors.Wrap(err, "create snapshot")
+	}
+
+	snap.ShardingState = shardingState
+	snap.Schema = schema
+	snap.CompletedAt = time.Now()
+
+	if err := snap.Write(); err != nil {
+		return nil, err
+	}
+
+	return snap, nil
+}
+
+// ReleaseSnapshot marks the specified snapshot as inactive and restarts all
+// async background and maintenance processes. It errors if the snapshot does not exist
+// or is already inactive.
+func (i *Index) ReleaseSnapshot(ctx context.Context, id string) error {
+	i.snapshotState = snapshots.State{InProgress: false}
+	return nil
+}
+
+func (i *Index) initSnapshot(id string) error {
```

Comment:
create snapshot will be started by http request, so multiple calls can happen at once. would be safer if snapshotState update was protected by some lock to prevent race issues

Reply:
Agreed, we need to guarantee two things, I believe:

- Only one snapshot request is occurring at the same time. (i.e. mutual exclusion for anything releated to snapshots)
-  During the creation of a snapshot, write requests must wait, i.e. something like multi-reader mutual exclusion with all "write requests" where a regular write represents a Reader (there can be many as long as no snapshot is going on) in an RWMutex and a Snapshot creation represents a Writer (there can only ever be a single one, and no other requests)

Reply:
@aliszka @etiennedi I think the locking mechanism should exist on whichever entity is responsible for calling `*Index.CreateSnapshot`, no?

Reply:
@aliszka ah, I misread your comment. Agreed, we should lock snapshot state to avoid a race.